lib/datadog/core/remote/component.rb

```diff
@@ -23,27 +23,71 @@ def initialize(settings, agent_settings)
           capabilities = Client::Capabilities.new(settings)
 
           @client = Client.new(transport_v7, capabilities)
-          @worker = Worker.new(interval: settings.remote.poll_interval_seconds) { @client.sync }
-        end
+          @worker = Worker.new(interval: settings.remote.poll_interval_seconds) do
+            begin
+              @client.sync
+            rescue StandardError => e
+              Datadog.logger.error do
+                "remote worker error: #{e.class.name} #{e.message} location: #{Array(e.backtrace).first}"
+              end
 
-        def barrier(kind)
-          return if @worker.nil?
+              # client state is unknown, state might be corrupted
+              @client = Client.new(transport_v7, capabilities)
+
+              # TODO: bail out if too many errors?
+            end
+
+            @barrier.lift
+          end
 
-          # Make it start on demand (for now)
-          @worker.start
+          @barrier = Barrier.new { @worker.start unless @worker.nil? }
+        end
 
+        def barrier(kind)
           case kind
           when :once
-            # TODO: block until first update has been received
+            @barrier.wait_once
           when :next
-            # TODO: block until next update has been received
+            @barrier.wait_next
           end
         end
 
         def shutdown!
           @worker.stop unless @worker.nil?
         end
 
+        class Barrier
+          def initialize(&block)
+            @block = block
+            @once = false
+
+            @mutex = Mutex.new
+            @condition = ConditionVariable.new
+          end
+
+          def wait_once
+            return if @once
+
+            wait_next
+          end
+
+          def wait_next
+            return unless @block.call
+
+            @mutex.synchronize do
+              @condition.wait(@mutex)
+            end
+          end
+
+          def lift
+            @mutex.synchronize do
+              @once ||= true
+
+              @condition.broadcast
+            end
+          end
+        end
```

Comment:
Hey! I'm always curious when concurrency stuff shows up. May I suggest adding a few comments explaining what this is and how it's used?

In particular, I'm curious about the intended semantics, with e.g. calling the block outside of the lock, and checking `@once` outside as well, but having the condition variable and setting `@once` inside the lock.

Reply:
I will! And also figure out some better naming.

For context: I was trying to get the implementation right + a bunch of passing specs before diving into comments.

Reply:
> Thread 1 finally calls wait_next

You mean, because it it scheduled back in, resuming `wait_once` after bullet point 1?

> Can it get stuck there waiting for a lift that already happened?

Well if there's only one `lift`, yes. But the expectation is that `lift` is happening in a loop and gets called regularly. So worst case it gets next `lift`, which means `wait_once` waited for a `lift` _at least_ once (not _exactly_ once).

Practically, we're using `wait_once` so that we wait for a RC sync to happen at least once before proceeding.

Same goes for `wait_next`, it is meant to wait for a subsequent `lift`, not necessarily the _immediate_ next one (which is a fuzzy thing to define without additional synchronization in the Barrier caller).

That said, I'm fine with doing this instead:

```
          def wait_once
            @mutex.synchronize do
              return if @once

              @condition.wait(@mutex)
            end
          end
```

Which should cover your concern at that level.

Reply:
Done.

Reply:
> Well if there's only one lift, yes. But the expectation is that lift is happening in a loop and gets called regularly. So worst case it gets next lift, which means wait_once waited for a lift at least once (not exactly once).

I think we're really missing some higher-level documentation on how this is intended to get used and all of the assumptions being made. This expectation you mentioned was not clear to me at all; and even after you mentioned it, it's still not very clear to me what "gets called regularly" means.

Looking at the code, it looks like this `barrier` gets called from a regular request thread, so does missing a wake-up mean that it can hang until the next remote configuration upgrade comes in?

Writing comments can come across as really dry so I just want to clarify that I'm trying to help :smile: Getting this mechanism wrong I think can both have high impact (e.g. disrupting user requests) as well as be very hard to debug, so I'm asking all of these things to help review this code.

Reply:
> Looking at the code, it looks like this barrier gets called from a regular request thread, so does missing a wake-up mean that it can hang until the next remote configuration upgrade comes in?

With `wait_next`, yes, but the request barrier is using `wait_once`, so the blocking is only happening until there's at least one sync attempt, which happens "right away" because the RC worker is lazily-started immediately (or rather, as immediately as thread scheduling allows it)

Reply:
>  I just want to clarify that I'm trying to help

I did not take it any other way! ;)

Reply:
So when does `wait_next` get used instead? Is it only for testing?